avro_schema/utils.lua

```diff
@@ -53,8 +53,137 @@ function has_only(t, key)
     return fst_key == key and snd_key == nil
 end
 
+local function map(func, tab)
+    local res = {}
+    for k, v in pairs(tab) do
+        res[k] = func(k, v)
+    end
+    return res
+end
+
+-- Fast stack implementation.
+-- This stack can store several value in each frame.
+-- instead of storing tables for each frame, this stack maintains
+-- several arrays and stores different keys to different values.
+--
+-- E.g. init_fstack({'schema', 'data'}) would create a table like this:
+-- {
+--      schema = {}, -- array which stores schema values
+--      set_schema = function (schema),
+--      get_schema = function (),
+--      data = {},   -- array which stores data values
+--      set_data = function (data),
+--      get_data = function (),
+--      len = 0,     -- current size of the array
+--      push = function (schema, data),
+--      pop  = function () -> { schema = schema[i], data = data[i] }
+--      get  = function (i) -> { schema = schema[i], data = data[i] }
+--      remove_last = function (),
+--      clear = function (),
+-- }
+local function init_fstack(keys)
+    assert(table_contains(keys, 'push') == false)
+    assert(table_contains(keys, 'pop') == false)
+    assert(table_contains(keys, 'remove_last') == false)
+    assert(table_contains(keys, 'get') == false)
+    assert(table_contains(keys, 'clear') == false)
+    local push_str = [[
+    return function(stack, VARS)
+        stack.len = stack.len + 1
+        SET_VALS
+    end]]
+    push_str = push_str:gsub('VARS',
+        table.concat(
+            map(function (i, _) return 'v' .. i end, keys),
+            ',')
+        )
+    push_str = push_str:gsub('SET_VALS',
+        table.concat(
+            map(function (i, key)
+                return ('stack.%s[stack.len] = v%d'):format(key, i)
+            end, keys),
+            '\n')
+        )
+    local pop_str = [[
+    return function (stack)
+        local res = stack:get(stack.len)
+        stack:remove_last()
+        return res
+    end
+    ]]
+    local remove_last_str = [[
+    return function (stack)
+        stack.len = stack.len - 1
```

Comment:
I think we should free pointers here. At least we should measure whether perf will be degraded and if so properly comment this place.

Reply:
~10% degradation on benchmark tests.
Comment added.